lib/ace/ext/papeeria/katex-previewer.coffee

```diff
@@ -50,43 +55,136 @@ define((require, exports, module) ->
       )
       return
 
-    callbackHidePopover = ->
-      popoverHandler.hide($("#formula"))
-      editor.off("changeSelection", callbackHidePopover)
-      editor.session.off("changeScrollTop", callbackHidePopover)
-      editor.session.off("changeScrollLeft", callbackHidePopover)
-      return
+    jqEditorContainer = $(editor.container)
+    jqFormula = -> $("#formula")
+
+    ch = contextHandler = {
+      contextPreviewExists: false
+      updateDelay: 1000
+
+      removeRegex: /\\end\{equation\}|\\begin\{equation\}|\\label\{[^\}]*\}/g
 
-    onLoaded = ->
-      try
-        content = katex.renderToString(
-          editor.getSelectedText(),
+      getEquationRange: (cursorRow) ->
+        i = cursorRow
+        while latexContextParser.getContext(editor.session, i - 1) == "equation"
+          i -= 1
+        start = i
+        while latexContextParser.getContext(editor.session, i + 1) == "equation"
+          i += 1
+        end = i
+        return [start, end]
+
+      getWholeEquation: (start, end) ->
+        editor.session.getLines(start, end).join(" ").replace(ch.removeRegex, "")
+
+      getPopoverPosition: (row) -> {
+          top: "#{editor.renderer.textToScreenCoordinates(row + 2, 1).pageY}px"
+          left: "#{jqEditorContainer.position().left}px"
+        }
+
+      getCurrentFormula: ->
+        katex.renderToString(
+          ch.getWholeEquation(ch.curStart, ch.curEnd),
           {displayMode: true}
         )
-      catch e
-        content = e
-      finally
-        popoverHandler.show($("#formula"), content)
-        editor.on("changeSelection", callbackHidePopover)
-        editor.session.on("changeScrollTop", callbackHidePopover)
-        editor.session.on("changeScrollLeft", callbackHidePopover)
-        return
 
-    callback = (editor) -> createPopover(editor)
+      initPopover: ->
+        cursorRow = editor.getCursorPosition().row
+        [ch.curStart, ch.curEnd] = ch.getEquationRange(cursorRow)
+        popoverPosition = ch.getPopoverPosition(ch.curEnd)
+        try
+          content = ch.getCurrentFormula()
+        catch e
+          content = e
+        finally
+          popoverHandler.show(jqFormula(), content, popoverPosition)
 
-    destroyPopover = -> popoverHandler.hide($("#formula"))
+      updatePopover: ->
+        try
+          content = ch.getCurrentFormula()
+        catch e
+          content = e
```

Comment:
Okay, I'l add TODO. I'm working right now on filtering out things like `begin{equation}` using tokenizer instead of regex, so I guess that would be the next thing.